File: ios/Classes/MergeParam.swift
import Flutter
import Foundation
import UIKit

struct MergeParam {
    let color: UIColor?
    let width: CGFloat
    let height: CGFloat
    let format: Int
    let quality: Int
    let imageParams: [ImageParam]

    init?(from dict: [String: Any]) {
        // Handle both Int and Double from Dart
        guard let width = (dict["width"] as? NSNumber)?.doubleValue,
              let height = (dict["height"] as? NSNumber)?.doubleValue,
              let format = dict["format"] as? Int,
              let quality = dict["quality"] as? Int,
              let images = dict["imageParams"] as? [[String: Any]] else {
            return nil
        }

        self.width = CGFloat(width)
        self.height = CGFloat(height)
        self.format = format
        self.quality = min(max(quality, 0), 100) // Clamp quality 0-100
        self.imageParams = images.compactMap { ImageParam(from: $0) }

        // Validate image params exist
        if self.imageParams.isEmpty {
            return nil
        }

        // Safer color parsing
        if let rgba = dict["color"] as? [NSNumber], rgba.count == 4 {
            let components = rgba.map { CGFloat($0.floatValue) / 255.0 }
            self.color = UIColor(
                red: components[1],
                green: components[2],
                blue: components[3],
                alpha: components[0]
            )
        } else {
            self.color = nil
        }
    }
}

struct ImageParam {
    let image: Data
    let dx: CGFloat
    let dy: CGFloat
    let width: CGFloat
    let height: CGFloat

    init?(from dict: [String: Any]) {
        guard let image = dict["image"] as? FlutterStandardTypedData,
              let dx = (dict["dx"] as? NSNumber)?.doubleValue,
              let dy = (dict["dy"] as? NSNumber)?.doubleValue,
              let width = (dict["width"] as? NSNumber)?.doubleValue,
              let height = (dict["height"] as? NSNumber)?.doubleValue else {
            return nil
        }

        self.image = image.data
        self.dx = CGFloat(dx)
        self.dy = CGFloat(dy)
        self.width = CGFloat(width)
        self.height = CGFloat(height)

        // Validate image data
        if self.image.isEmpty {
            return nil
        }
    }
}

File: lib/src/merge_param.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:widget_screenshot_plus/src/widget_screenshot_plus.dart';

class MergeParam {
  final Color? color;
  final Size size;
  final ShotFormat format;
  final int quality;
  final List<ImageParam> imageParams;

  MergeParam({
    this.color,
    required this.size,
    required this.format,
    required this.quality,
    required this.imageParams,
  });

  Map<String, dynamic> toJson() => {
    if (color != null)
      "color": [
        (color!.a * 255.0).round(),
        (color!.r * 255.0).round(),
        (color!.g * 255.0).round(),
        (color!.b * 255.0).round(),
      ],

    "width": size.width,
    "height": size.height,
    "format": format == ShotFormat.png ? 0 : 1,
    "quality": quality,
    "imageParams": imageParams.map((e) => e.toJson()).toList(),
  };
}

class ImageParam {
  final Uint8List image;
  final Offset offset;
  final Size size;

  ImageParam({required this.image, required this.offset, required this.size});

  ImageParam.start(Uint8List image, Size size)
    : this(image: image, offset: const Offset(-1, -1), size: size);

  ImageParam.end(Uint8List image, Size size)
    : this(image: image, offset: const Offset(-2, -2), size: size);

  Map<String, dynamic> toJson() => {
    "image": image,
    "dx": offset.dx,
    "dy": offset.dy,
    "width": size.width,
    "height": size.height,
  };
}
